translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,355 @@
+// limitations under the License.
+
+package metrics
+
+import (
+	"sort"
+
+	ocmetrics "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	otlpcommon "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	translatorcommon "github.com/open-telemetry/opentelemetry-collector/translator/common"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := translatorcommon.ResourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	labelKeys := allLabelKeys(metric)
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor, labelKeys),
+		Timeseries:       dataPointsToTimeseries(metric, labelKeys),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor, labelKeys []*ocmetrics.LabelKey) *ocmetrics.MetricDescriptor {
+	if descriptor == nil {
+		return nil
+	}
+
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        descriptorTypeToOC(descriptor.Type),
+		LabelKeys:   labelKeys,
+	}
+}
+
+func allLabelKeys(metric *otlpmetrics.Metric) []*ocmetrics.LabelKey {
+	// First, collect a set of unique keys
+	uniqueKeys := make(map[string]struct{}, 0)
+	for _, point := range metric.Int64Datapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.DoubleDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.HistogramDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.SummaryDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+
+	// Sort keys (this is actually optional, just for stable order)
+	rawKeys := make([]string, 0, len(uniqueKeys))
+	for key := range uniqueKeys {
+		rawKeys = append(rawKeys, key)
+	}
+	sort.Strings(rawKeys)
+
+	// Construct a list of label keys
+	// Note: Label values will have to match keys by index
+	labelKeys := make([]*ocmetrics.LabelKey, 0, len(rawKeys))
+	for _, key := range rawKeys {
+		labelKeys = append(labelKeys, &ocmetrics.LabelKey{
+			Key: key,
+		})
+	}
```

Comment:
Instead of allocating the ocmetrics.LabelKey one by one, you can also have a slice of ocmetrics.LabelKey and point to that. This reduce the number of allocations.

Reply:
@tigrannajaryan previously mentioned that it's fine to not optimize allocations in this translator since it's temporary.

Reply:
I'd rather have this done quickly, then done optimally. I consider this just a temporary implementation, which we will hopefully retire in a few week. We still need it to be correct though, so no shortcuts here :-)